RockLib.Messaging.NamedPipes/NamedPipeSender.cs

```diff
@@ -61,75 +70,53 @@ public Task SendAsync(SenderMessage message, CancellationToken cancellationToken
         /// </summary>
+#pragma warning disable CA1063 // Implement IDisposable Correctly
+#pragma warning disable CA1816 // Dispose methods should call SuppressFinalize
         public void Dispose()
+#pragma warning restore CA1816 // Dispose methods should call SuppressFinalize
+#pragma warning restore CA1063 // Implement IDisposable Correctly
         {
+            _senderCancellation.Cancel();
+
             _workItems.CompleteAdding();
-            _runThread.Join();
+            _workItems.Dispose();
+
+            _senderCancellation.Dispose();
+
+            if(_sender.IsCompleted)
+            {
+                _sender.Dispose();
+            }
         }
 
-        private void Run()
+        private void Run(CancellationToken token)
         {
-            foreach (var workItem in _workItems.GetConsumingEnumerable())
+            try
             {
-                if (workItem.CancellationToken.IsCancellationRequested)
+                foreach (var workItem in _workItems.GetConsumingEnumerable(token))
                 {
-                    workItem.Completion.SetCanceled();
-                    continue;
-                }
+                    using var pipe = new NamedPipeClientStream(".", PipeName, PipeDirection.Out, PipeOptions.Asynchronous);
+                    pipe.Connect(0);
 
-                try
-                {
-                    using (var pipe = new NamedPipeClientStream(".", PipeName, PipeDirection.Out, PipeOptions.Asynchronous))
-                    {
-                        try
-                        {
-                            pipe.Connect(0);
-                        }
-                        catch (TimeoutException ex)
-                        {
-                            workItem.Completion.SetException(ex);
-                            continue;
-                        }
-
-                        if (workItem.CancellationToken.IsCancellationRequested)
-                        {
-                            workItem.Completion.SetCanceled();
-                            continue;
-                        }
-
-                        using (var writer = new StreamWriter(pipe))
-                        {
-                            writer.WriteLine(workItem.Message);
-                        }
-
-                        workItem.Completion.SetResult(true);
-                    }
-                }
-                catch (Exception ex)
-                {
-                    workItem.Completion.SetException(ex);
+                    using var writer = new StreamWriter(pipe);
+                    writer.WriteLine(workItem);
                 }
             }
-        }
-
-        private struct WorkItem
-        {
-            public string Message;
-            public TaskCompletionSource<bool> Completion;
-            public CancellationToken CancellationToken;
+            catch (OperationCanceledException) { }
```

Comment:
is it weird that we are catching the exception but not doing anything with it?

Reply:
This is one of the few acceptable cases. If a `CancellationToken` has been canceled, and someone calls [`ThrowIfCancellationRequested`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.cancellationtoken.throwifcancellationrequested?view=net-6.0#system-threading-cancellationtoken-throwifcancellationrequested), then it'll throw that exception, but that's expected. In that case, you just catch and move on.